q2_metadata/normalization/_norm_rules.py

```diff
@@ -0,0 +1,304 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import yaml
+from glob import glob
+import pandas as pd
+from os.path import basename, isdir, splitext
+
+from q2_metadata.normalization._norm_messages import WarningsCollection, ErrorsCollection
+from q2_metadata.normalization._norm_check_rules import check_rule
+
+
+class Rules(object):
+    """Collect the rules of one current variable.
```

Comment:
Any chance you could move the descriptions of the rules in this docstring to a module-wide docstring?

Reply:
will do.

Reply:
like this?

```
# ----------------------------------------------------------------------------
# Copyright (c) 2016-2020, QIIME 2 development team.
#
# Distributed under the terms of the Modified BSD License.
#
# The full license is in the file LICENSE, distributed with this software.
# ----------------------------------------------------------------------------

"""This module contain the main classes for the metadata normalization
plugin in QIIME2.

Glossary:
    variable        :   Name of a metadata variable supposed to occur in
                        the metadata table header.
    variable_rules  :   Set of rules associated with a variable and described
                        in the per-variable .yml files.
    rules           :   collection of all the rules that will apply on all
                        variables. The rules are parsed, checked and collected
                        into this structured object before being applied.
    rule            :   One rule that is one of "expected", "ontology",
                        "remap", "validation", "normalization", "blank",
                        "missing" or "format". Each of these rules has a
                        specific format that is checked (module
                        _norm_check_rules.py).

Rules formatting (.yml files):
    remap           :   Map the variables values by replacing all instances
                        of a given value by another one value.
        Example:
            ```
            remap:
              US: USA
            ```
    validation      :   Validate the variables values by replacing by the
                        missing value every sample entry that for the current
                        variable has in one or multiple other variable(s) a
                        null or missing value for that sample.
        Example:
            ```
            validation:
              force_to_blank_if:
                is null:
                - variable
            ```
    normalization   :   Normalize the variables values by replacing the values
                        outside the a range by a given replacement value.
        Example:
            ```
            normalization:
                gated_value: nan
                maximum: 100
                minimum: 10
            ```
    missing         :   Check that missing values are one of the terms reserved
                        for this instance.
        Example:
            ```
            missing:
                - not applicable
                - not collected
                - not provided
                - restricted access
            ```
    format          :   Check that the passed column is in the correct format.
        Example:
            ```
            format:
              - bool
              - float
              - int
              - str
            ```

    expected        :   Check that the variable values are one of the passed
                        expected values (or the missing and blanks).
        Example:
            ```
            expected:
                - "Yes"
                - "No"
            ```
    ontology        :   Name of the ontology to look up for control vocabulary.
        Example:
            ```
            ontology: Gazetteer
            ```
    blank           :   Check that blank values are one of the terms reserved
                        for this instance.
        Example:
            ```
            missing:
                - not applicable
                - not collected
                - not provided
                - restricted access
            ```
"""

import yaml
from glob import glob
import ...
```